group theory: more efficient Sylow subgroups for Sym and Alt groups

valglad opened the pull request:
This PR adds a separate method `_sylow_alt_sym` for computation of Sylow subgroups of symmetric and alternating groups, to be used as part of the main `sylow_subgroup` method. This is much more efficient than the general approach, so that for example, a 3-Sylow subgroup of `SymmetricGroup(100)` can be found within a hundredth of a second compared to more than a minute that it currently takes (probably much more, I stopped waiting).

Additionally, `is_alt_sym` is slightly modified to handle the case of permutation groups of degree less than 8.

valglad committed:
_sylow_alt_sym

valglad committed:
Merge branch 'master' into alt_sym_sylow

valglad committed:
edit tests and docstring

jksuom commented on sympy/combinatorics/perm_groups.py:
```diff
@@ -3556,6 +3561,101 @@ def _p_elements_group(G, p):
+        # subgroup on [0..p**(i-1)-1]
+        for i in range(1, power+1):
+            gen = Permutation([(j + p**(i-1)) % p**i for j in range(p**i)])
+            if alt and gen == Permutation(0, 1):
```
Could the test be `if i == 1 and alt:` (before `gen` is created)?

valglad replied:
Yes, that would be better.

valglad committed:
replace reduce and alt test

jksuom commented on sympy/combinatorics/perm_groups.py:
```diff
@@ -3556,6 +3563,101 @@ def _p_elements_group(G, p):
+                    # shift the generator to the start of the
+                    # partition part
+                    gen = shift*gen*shift
+                    gens.append(gen)
```
It seems that this is not necessary when `start == 0` (no shift).

valglad committed:
better generators for alt and change alt check

This commit rearranges and modifies the code related to generators
of alternating groups so that they are consistent with those
desribed in the docstring. Also, the check for whether a group is
alternating is replaced by considering its order: checking if (0 1)
is contained in the group takes far too long for large degrees.

valglad committed:
change the alternating group check

valglad committed:
change generator order for alt
